app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php

```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
-            ScopeInterface::SCOPE_STORE
-        )) {
-            $cssMatches = [];
-            // add link rel preload to style sheets
-            $content = preg_replace_callback(
-                '@<link\b.*?rel=("|\')stylesheet\1.*?/>@',
-                function ($matches) use (&$cssMatches) {
-                    $cssMatches[] = $matches[0];
-                    preg_match('@href=("|\')(.*?)\1@', $matches[0], $hrefAttribute);
-                    $href = $hrefAttribute[2];
-                    if (preg_match('@media=("|\')(.*?)\1@', $matches[0], $mediaAttribute)) {
-                        $media = $mediaAttribute[2];
-                    }
-                    $media = $media ?? 'all';
-                    $loadCssAsync = sprintf(
-                        '<link rel="preload" as="style" media="%s"' .
-                         ' onload="this.onload=null;this.rel=\'stylesheet\'"' .
-                        ' href="%s" />',
-                        $media,
-                        $href
-                    );
-
-                    return $loadCssAsync;
-                },
-                $content
-            );
+        $content = (string)$httpResponse->getContent();
+        $headCloseTag = '</head>';
+
+        $headEndTagFound = strpos($content, $headCloseTag) !== false;
+
+        if ($headEndTagFound) {
+            $styles = $this->extractLinkTags($content);
+            if ($styles) {
+                $newHeadEndTagPosition = strrpos($content, $headCloseTag);
+                $content = substr_replace($content, $styles . "\n", $newHeadEndTagPosition, 0);
+                $httpResponse->setContent($content);
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Extracts link tags found in given content.
+     *
+     * @param string $content
+     */
+    private function extractLinkTags(string &$content): string
+    {
+        $styles = '';
+        $styleOpen = '<link';
+        $styleClose = '>';
+        $styleOpenPos = strpos($content, $styleOpen);
 
-            if (!empty($cssMatches)) {
-                $content = str_replace('</body', implode("\n", $cssMatches) . "\n</body", $content);
-                $subject->setContent($content);
+        while ($styleOpenPos !== false) {
+            $styleClosePos = strpos($content, $styleClose, $styleOpenPos);
+            $style = substr($content, $styleOpenPos, $styleClosePos - $styleOpenPos + strlen($styleClose));
+
+            if (!preg_match('@rel=["\']stylesheet["\']@', $style)) {
+                // Link is not a stylesheet, search for another one after it.
+                $styleOpenPos = strpos($content, $styleOpen, $styleClosePos);
+                continue;
+            }
+            // Remove the link from HTML to add it before </head> tag later.
+            $content = str_replace($style, '', $content);
+
+            preg_match('@href=("|\')(.*?)\1@', $style, $hrefAttribute);
+            $href = $hrefAttribute[2];
```

Comment:
I'm not sure about this code behaviour when there's invalid syntax provided.
Would you mind adding extra condition:
```suggestion
            if (!preg_match('@href=("|\')(.*?)\1@', $style, $hrefAttribute)) {
                throw new \Exception('Invalid <style> syntax provided');
            }
            $href = $hrefAttribute[2];
```

or something similar.